bayes_kit/drghmc.py

```diff
@@ -0,0 +1,437 @@
+            rho_mid += stepsize * np.multiply(self._metric, grad).squeeze()
+            theta += stepsize * rho_mid
+            _, grad = self._model.log_density_gradient(theta)
+        rho = rho_mid + 0.5 * stepsize * np.multiply(self._metric, grad).squeeze()
+        return (theta, rho)
+
+    def _prob_of_delayed_rejection(self, hastings: float) -> float:
+        """Log probability of making another proposal upon rejection.
+
+        To reduce average cost per iteration for DRHMC, make the delayed rejections
+        probabilistic, such that a subsequent proposal is not mandatory upon rejection.
+
+        Probabilistic delayed rejection proposes a new sample with some probability
+        dependent on where we are in the distribution. If a proposal has low acceptance
+        probability and was rejected, propose another sample; if a proposal has high
+        acceptance probablity but was rejected by chance, do not propose another
+        sample. More details found in section 3.2 of Modi et al.
+
+        A heuristic proposal probability that achieve this is eqn. 31 in Modi et al:
+        the probability of making proposal k+1, after rejecting proposal k, is the
+        probability of rejecting proposal k. The proposal probability of making the
+        first proposal, however, is always 1.
+
+        This proposal probability modifies the proposal kernel, which alters detailed
+        balance. Thus the probability of accepting a proposal is multiplied by the
+        hasting term of the propopsed sample divided by the hastings term of the
+        current sample.
+
+        Args:
+            hastings: log probability of rejecting all previous proposals
+
+        Returns:
+            float: log probability of making another proposal
+        """
+        return hastings if self._probabilistic else 0.0
+
+    def sample(self) -> DrawAndLogP:
+        """Sample from target distribution with DrGhmcDiag sampler.
+
+        From current sample (theta, rho), propose new sample (theta_prop, rho_prop)
+        and compute its acceptance probability. If accepted, return the new sample and
+        its log density; if rejected, propose a new sample and repeat. Stop when the
+        maximum number of proposals is reached or when probabilistic delayed rejection
+        terminates early.
+
+        Flip momemntum after each proposal to ensure detailed balance. Flip momentum
+        before return statement for momentum persistence in Generalized HMC.
+
+        Returns:
+            DrawAndLogP: Tuple of (sample, log density)
+        """
+        self._rho = self._rng.normal(
+            loc=self._rho * np.sqrt(1 - self._dampening),
+            scale=self._dampening,
+            size=self._dim,
+        )
+        logp_cur = self.joint_logp(self._theta, self._rho)
+        hastings_cur, reject_logp = 0.0, 0.0
+
+        for k in range(self._num_proposals):
+            pdr = self._prob_of_delayed_rejection(reject_logp)
```

Comment:
I think this logic is wrong.  Doesn't `_prob_of_delayed_rejection` always return `False` if there `self.probabilistic = False`?  Have you tested this in the situation with `self.probabilistic = False`?

Reply:
No.

If `self.probabilistic = False`, then `_prob_of_delayed_rejection` always returns 0 on the log scale, or 1 on the linear scale. Thus the probability of "retrying" another proposal upon rejection is always 1. 

@bob-carpenter Am I missing something here? Did you see this differently?

Reply:
I believe the `_prob_of_delayed_rejection` function would not always return false. 

Perhaps this was unclear because the `_prob_of_delayed rejection` function is on the log scale? If so, see https://github.com/flatironinstitute/bayes-kit/pull/39#discussion_r1260302454.